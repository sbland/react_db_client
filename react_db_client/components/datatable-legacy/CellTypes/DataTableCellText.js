import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';

const DataTableCellText = ({
  columnData: { type },
  cellData,
  updateData,
  acceptValue,
  resetValue,
  focused,
  editMode,
}) => {
  const refText = useRef(null);
  const refArea = useRef(null);
  // const [ignoreNextBlur, setIgnoreNextBlur] = useState(false);

  const handleInputChange = (e) => {
    updateData(e.target.value);
  };

  useEffect(() => {
    if (focused && editMode && type === 'text') {
      refText.current.focus();
      refText.current.select();
    }
    if (focused && editMode && type === 'textLong') {
      refArea.current.focus();
      refArea.current.select();
    }
  }, [focused, refText, refArea, type, editMode]);

  const acceptValueLocal = () => {
    acceptValue(cellData);
  };

  const rejectValue = () => {
    resetValue();
  };

  const onKeyPress = (e) => {
    // setIgnoreNextBlur(true);
    if (e.key === 'Enter' || e.key === 'Tab') {
      e.preventDefault();
      acceptValueLocal();
    }
    if (e.key === 'Escape') {
      e.preventDefault();
      rejectValue();
    }
  };

  const onBlur = () => {
    // if (!ignoreNextBlur) acceptValueLocal();
    acceptValueLocal();
    // setIgnoreNextBlur(false);
  };

  const classNames = [
    type === 'textLong' ? 'dataTableCellData-textLong' : 'dataTableCellData-text',
  ].join(' ');

  const showTextEditor = focused && editMode && type === 'text';
  const showTextAreaEditor = focused && editMode && type === 'textLong';
  return (
    <div className={`dataTableCellData ${classNames}`}>
      <input
        style={{
          display: showTextEditor ? 'block' : 'none',
        }}
        className="cellInput-text"
        ref={refText}
        type="text"
        onChange={handleInputChange}
        value={cellData || ''}
        onBlur={onBlur}
        onKeyDown={onKeyPress}
      />
      {/* <input
        style={{
          display: showTextAreaEditor ? 'block' : 'none',
        }}
        className="cellInput-textarea"
        ref={refArea}
        type="text"
        onChange={handleInputChange}
        value={cellData || ''}
        onBlur={acceptValueLocal}
        onKeyDown={onKeyPress}
      /> */}
      {/* // Disabled text area as does not fit well */}
      <textarea
        style={{
          display: showTextAreaEditor ? 'block' : 'none',
          position: 'absolute',
          width: '100%',
          minHeight: '5rem',
          minWidth: '20rem',
          left: 0,
          top: 0,
          whiteSpace: 'normal',
          resize: 'none',
          zIndex: 10,
          overflow: 'hidden',
        }}
        className="cellInput-textarea"
        ref={refArea}
        type="text"
        onChange={handleInputChange}
        value={cellData || ''}
        onBlur={onBlur}
        onKeyDown={onKeyPress}
        wrap="hard"
        rows="10"
        cols="20"
      />
      {(!editMode || !focused) && (
        <div className={`dataTableCellData_text ${classNames}`}>{cellData}</div>
      )}
    </div>
  );
};

DataTableCellText.propTypes = {
  cellData: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  columnData: PropTypes.shape({
    defaultValue: PropTypes.string,
    type: PropTypes.oneOf(['text', 'textLong']).isRequired,
  }).isRequired,
  updateData: PropTypes.func.isRequired,
  acceptValue: PropTypes.func.isRequired,
  resetValue: PropTypes.func.isRequired,
  focused: PropTypes.bool.isRequired,
  editMode: PropTypes.bool.isRequired,
};

DataTableCellText.defaultProps = {
  cellData: '',
};

export default DataTableCellText;
