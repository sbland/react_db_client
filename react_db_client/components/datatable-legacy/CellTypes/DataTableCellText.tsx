import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { DefaultCellInnerStyle } from './style';
import { ICellProps, IHeadingText } from '../lib';

export interface IDataTableCellTextProps extends ICellProps<IHeadingText> {
  cellData: string;
}

export const DataTableCellText = ({
  columnData: { type},
  rowId,
  columnId,
  cellData,
  updateData,
  acceptValue,
  resetValue,
  focused,
  editMode,
}: IDataTableCellTextProps) => {
  const refText = useRef<HTMLInputElement>(null);
  const refArea = useRef<HTMLTextAreaElement>(null);
  // const [ignoreNextBlur, setIgnoreNextBlur] = useState(false);

  const handleInputChange = (e) => {
    updateData(e.target.value, rowId, columnId);
  };

  useEffect(() => {
    if (focused && editMode && type === 'text') {
      refText.current?.focus();
      refText.current?.select();
    }
    if (focused && editMode && type === 'textLong') {
      refArea.current?.focus();
      refArea.current?.select();
    }
  }, [focused, refText, refArea, type, editMode]);

  const acceptValueLocal = () => {
    acceptValue(cellData);
  };

  const rejectValue = () => {
    resetValue();
  };

  const onKeyPress = (e) => {
    // setIgnoreNextBlur(true);
    if (e.key === 'Enter' || e.key === 'Tab') {
      e.preventDefault();
      acceptValueLocal();
    }
    if (e.key === 'Escape') {
      e.preventDefault();
      rejectValue();
    }
  };

  const onBlur = () => {
    // if (!ignoreNextBlur) acceptValueLocal();
    if (focused) {
      acceptValueLocal();
    }
    // setIgnoreNextBlur(false);
  };

  const classNames = [
    type === 'textLong' ? 'dataTableCellData-textLong' : 'dataTableCellData-text',
  ].join(' ');

  const isEditing = focused && editMode;
  return (
    <DefaultCellInnerStyle className={`dataTableCellData ${classNames}`}>
      {type === 'text' && (
        <input
          style={{
            display: isEditing ? 'block' : 'none',
          }}
          className="cellInput-text"
          ref={refText}
          type="text"
          onChange={handleInputChange}
          value={cellData || ''}
          onBlur={onBlur}
          onKeyDown={onKeyPress}
        />
      )}
      {/* <input
        style={{
          display: showTextAreaEditor ? 'block' : 'none',
        }}
        className="cellInput-textarea"
        ref={refArea}
        type="text"
        onChange={handleInputChange}
        value={cellData || ''}
        onBlur={acceptValueLocal}
        onKeyDown={onKeyPress}
      /> */}
      {/* // Disabled text area as does not fit well */}
      {type === 'textLong' && (
        <textarea
          style={{
            display: isEditing ? 'block' : 'none',
            position: 'absolute',
            width: '100%',
            minHeight: '5rem',
            minWidth: '20rem',
            left: 0,
            top: 0,
            whiteSpace: 'normal',
            resize: 'none',
            zIndex: 10,
            overflow: 'hidden',
          }}
          className="cellInput-textarea"
          ref={refArea}
          onChange={handleInputChange}
          value={cellData || ''}
          onBlur={onBlur}
          onKeyDown={onKeyPress}
          wrap="hard"
          rows={10}
          cols={20}
        />
      )}
      {!isEditing && <div className={`dataTableCellData_text ${classNames}`}>{cellData}</div>}
    </DefaultCellInnerStyle>
  );
};

DataTableCellText.propTypes = {
  cellData: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  columnData: PropTypes.shape({
    defaultValue: PropTypes.string,
    type: PropTypes.oneOf(['text', 'textLong']).isRequired,
  }).isRequired,
  updateData: PropTypes.func.isRequired,
  acceptValue: PropTypes.func.isRequired,
  resetValue: PropTypes.func.isRequired,
  focused: PropTypes.bool.isRequired,
  editMode: PropTypes.bool.isRequired,
};

DataTableCellText.defaultProps = {
  cellData: '',
};

export default DataTableCellText;
