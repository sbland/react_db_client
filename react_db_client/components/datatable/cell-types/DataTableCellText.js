import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import {
  DefaultCellInnerStyle,
  DefaultInputStyle,
  DefaultTextAreaStyle,
  Text,
  FocusedText,
  FocusedTextLong,
} from './style';

export const DataTableCellText = ({
  columnData: { type },
  cellData,
  updateData,
  acceptValue,
  resetValue,
  focused,
  editMode,
}) => {
  const refText = useRef(null);
  const refArea = useRef(null);
  const handleInputChange = (e) => {
    updateData(e.target.value);
  };

  useEffect(() => {
    if (focused && editMode && type === 'text') {
      refText.current.focus();
      refText.current.select();
    }
    if (focused && editMode && type === 'textLong') {
      refArea.current.focus();
      refArea.current.select();
    }
  }, [focused, refText, refArea, type, editMode]);

  const acceptValueLocal = () => {
    acceptValue(cellData);
  };

  const rejectValue = () => {
    console.log('RESETTING');
    resetValue();
  };

  const onKeyPress = (e) => {
    // setIgnoreNextBlur(true);
    if (e.key === 'Enter' || e.key === 'Tab') {
      e.preventDefault();
      acceptValueLocal();
    }
    if (e.key === 'Escape') {
      e.preventDefault();
      rejectValue();
    }
  };

  const onBlur = () => {
    // if (!ignoreNextBlur) acceptValueLocal();
    console.log('obblur');
    acceptValueLocal();
    // setIgnoreNextBlur(false);
  };

  const classNames = [
    type === 'textLong' ? 'dataTableCellData-textLong' : 'dataTableCellData-text',
  ].join(' ');

  const showTextEditor = focused && editMode && type === 'text';
  const showTextAreaEditor = focused && editMode && type === 'textLong';

  return (
    <DefaultCellInnerStyle className={`dataTableCellData ${classNames}`} focused={focused}>
      {type === 'text' && (
        <DefaultInputStyle
          style={{
            display: showTextEditor ? 'block' : 'none',
          }}
          className="cellInput-text"
          ref={refText}
          type="text"
          onChange={handleInputChange}
          value={cellData || ''}
          onBlur={onBlur}
          onKeyDown={onKeyPress}
        />
      )}
      {type === 'textLong' && (
        <DefaultTextAreaStyle
          style={{
            display: showTextAreaEditor ? 'block' : 'none',
          }}
          role="input"
          className="cellInput-textarea"
          ref={refArea}
          type="text"
          onChange={handleInputChange}
          value={cellData || ''}
          onBlur={onBlur}
          onKeyDown={onKeyPress}
          wrap="hard"
          rows="10"
          cols="20"
        />
      )}
      {!focused && <Text className={`dataTableCellData_text ${classNames}`}>{cellData}</Text>}
      {!editMode && focused && type === 'text' && (
        <FocusedText className={`dataTableCellData_text ${classNames}`}>{cellData}</FocusedText>
      )}
      {/* Make text resize to fit content on focus */}
      {!editMode && focused && type === 'textLong' && (
        <FocusedTextLong className={`dataTableCellData_text focusedtextCell ${classNames}`}>
          {cellData}
        </FocusedTextLong>
      )}
    </DefaultCellInnerStyle>
  );
};

DataTableCellText.propTypes = {
  cellData: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  columnData: PropTypes.shape({
    defaultValue: PropTypes.string,
    type: PropTypes.oneOf(['text', 'textLong']).isRequired,
  }).isRequired,
  updateData: PropTypes.func.isRequired,
  acceptValue: PropTypes.func.isRequired,
  resetValue: PropTypes.func.isRequired,
  focused: PropTypes.bool.isRequired,
  editMode: PropTypes.bool.isRequired,
};

DataTableCellText.defaultProps = {
  cellData: '',
};
