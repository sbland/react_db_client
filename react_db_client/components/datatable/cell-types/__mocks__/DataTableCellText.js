/* eslint-disable no-unused-vars */
/* eslint-disable arrow-body-style */
/* eslint-disable react/prop-types */
import React from 'react';

const DataTableCellText = ({
  cellData,
}) => {
  return (
    <div className="dataTableCellData text">
      {cellData}
    </div>
  );
};


export default DataTableCellText;
