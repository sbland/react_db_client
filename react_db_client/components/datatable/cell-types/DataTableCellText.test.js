import React from 'react';
import { mount } from 'enzyme';
import DataTableCellText from './DataTableCellText';

const updateData = jest.fn();
const acceptValue = jest.fn();
const resetValue = jest.fn();
const defaultProps = {
  columnData: {
    type: 'text',
  },
  cellData: 'Hello World',
  updateData,
  acceptValue,
  resetValue,
  rowData: {},
};

describe('DataTableCellType - text', () => {
  let component;

  const setEditMode = (c) => {
    c.setProps({ editMode: true, focused: true });
  };
  describe('Unit Testing', () => {
    beforeEach(() => {
      updateData.mockClear();
      acceptValue.mockClear();
      resetValue.mockClear();
      component = mount(<DataTableCellText {...defaultProps} />);
    });
    test('should show the cell data', () => {
      const tree = component.debug();
      expect(tree).toMatchSnapshot();
    });
    test('should show input when in edit mode', () => {
      let cellInput = component.find('.cellInput-text');
      expect(cellInput).toBeTruthy();
      expect(cellInput.props().style.display).toEqual('none');
      setEditMode(component);
      cellInput = component.find('.cellInput-text');
      expect(cellInput.props().style.display).toEqual('block');
    });
    test('should update data on cell data change', () => {
      setEditMode(component);
      const cellInput = component.find('.cellInput-text');
      const newValue = 'newentry';
      cellInput.simulate('change', { target: { value: newValue } });
      expect(updateData).toHaveBeenCalledWith(newValue);
    });
    test('should accept value on unfocus', () => {
      setEditMode(component);
      const cellInput = component.find('.cellInput-text');
      cellInput.simulate('blur');
      expect(acceptValue).toHaveBeenCalledWith(defaultProps.cellData);
    });
    test('should accept value on key down enter', () => {
      setEditMode(component);
      const cellInput = component.find('.cellInput-text');
      cellInput.simulate('keyDown', { key: 'Enter' });
      expect(acceptValue).toHaveBeenCalledWith(defaultProps.cellData);
    });
    test('should call reset if user presses escape', () => {
      setEditMode(component);
      const cellInput = component.find('.cellInput-text');
      const newValue = 'newentry';
      cellInput.simulate('change', { target: { value: newValue } });
      expect(updateData).toHaveBeenCalledWith(newValue);
      cellInput.simulate('keyDown', { key: 'Escape' });
      expect(resetValue).toHaveBeenCalledWith();
    });
  });
});
