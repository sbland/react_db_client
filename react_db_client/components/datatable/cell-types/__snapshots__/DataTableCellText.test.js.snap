// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableCellType - text Unit Testing should show the cell data 1`] = `
"<DataTableCellText columnData={{...}} cellData=\\"Hello World\\" updateData={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} acceptValue={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} resetValue={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [], instances: [], invocationCallOrder: [], results: [] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} rowData={{...}}>
  <div className=\\"dataTableCellData dataTableCellData-text\\">
    <input style={{...}} className=\\"cellInput-text\\" type=\\"text\\" onChange={[Function: handleInputChange]} value=\\"Hello World\\" onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyPress]} />
    <textarea style={{...}} className=\\"cellInput-textarea\\" type=\\"text\\" onChange={[Function: handleInputChange]} value=\\"Hello World\\" onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyPress]} wrap=\\"hard\\" rows=\\"10\\" cols=\\"20\\" />
    <div className=\\"dataTableCellData_text dataTableCellData-text\\">
      Hello World
    </div>
  </div>
</DataTableCellText>"
`;
